{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a4e7383a-d9b6-4083-aa7f-83b020bce89c",
   "metadata": {},
   "source": [
    "# OMGeEP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6389f2fc-0d41-40e1-b312-e31a75477049",
   "metadata": {},
   "source": [
    "# Download dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77733a3b-88f3-410d-96b9-3fccd0ecd6ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install lightgbm scikit-learn pandas numpy joblib\n",
    "# %pip install iterative-stratification\n",
    "# %pip install shap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c34a2a-3329-4861-b036-7f752d62577c",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fc9fce67-4bca-4453-a71b-b5dc12ca2d92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iterative-stratification not available — splits will be random.\n"
     ]
    }
   ],
   "source": [
    "import ast\n",
    "import re\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.multioutput import MultiOutputClassifier\n",
    "from sklearn.metrics import f1_score, hamming_loss, roc_auc_score, precision_recall_fscore_support\n",
    "from lightgbm import LGBMClassifier\n",
    "from joblib import Parallel, delayed\n",
    "\n",
    "# Optional iterative stratification (better for multilabel splits)\n",
    "try:\n",
    "    from iterstrat.ml_stratifiers import iterative_train_test_split\n",
    "    ITERATIVE_AVAILABLE = True\n",
    "except Exception:\n",
    "    ITERATIVE_AVAILABLE = False\n",
    "    print(\"iterative-stratification not available — splits will be random.\")\n",
    "\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "352e2e19-5434-471d-8ffe-b86b48f18620",
   "metadata": {},
   "source": [
    "# Global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7a8c096e-21cd-4b3a-a9f0-8a5c832fd418",
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_DIR = 'ifbdata/atlanteco_hack/OMGeEP/output_files'\n",
    "SAMPLE_ID_COL = 'sample_id'\n",
    "SITE_ID_COL = 'sampling_design_feature'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9ef7465-2280-4c0b-89f0-c716e8b3edb8",
   "metadata": {},
   "source": [
    "# General_Helper_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "7dbf5daa-163b-4652-b308-0247c0dbc191",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_most_variable_omics(df:pd.DataFrame, N:int=200, ID_col:str=SAMPLE_ID_COL) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    df: metagenomic gene df\n",
    "    N: Top N most variable genes\n",
    "    \"\"\"\n",
    "    copied_df = df.copy()\n",
    "    copied_df.set_index(ID_col,inplace=True)\n",
    "    gene_variance = copied_df.var(axis=1)\n",
    "    # Rank genes by variance\n",
    "    top_X = N   # choose the number of genes you want\n",
    "    top_genes = gene_variance.nlargest(top_X).index\n",
    "    copied_df_top = copied_df.loc[top_genes]\n",
    "    X = copied_df_top.T\n",
    "    return X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "699fdb0d-e11e-4feb-b314-a5d70608fde9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sample_id_mapping(file_path:str) -> pd.DataFrame:\n",
    "    mapping_file = pd.read_csv(file_path)\n",
    "    mapping_file = mapping_file[['sample-id', 'station-label']]\n",
    "    mapping_file.rename(columns={'sample-id':SAMPLE_ID_COL, 'station-label':SITE_ID_COL}, inplace=True)\n",
    "    return mapping_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c1bc26f9-0bb5-4a08-adbc-5bae1147f6f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sampling_design_feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA7950653</td>\n",
       "      <td>MMA-071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA7950654</td>\n",
       "      <td>MMA-071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA7950655</td>\n",
       "      <td>MMA-071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA7950656</td>\n",
       "      <td>MMA-071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA7918377</td>\n",
       "      <td>MMA-071</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id sampling_design_feature\n",
       "0  SAMEA7950653                 MMA-071\n",
       "1  SAMEA7950654                 MMA-071\n",
       "2  SAMEA7950655                 MMA-071\n",
       "3  SAMEA7950656                 MMA-071\n",
       "4  SAMEA7918377                 MMA-071"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_file = create_sample_id_mapping('Sample_ID_mapping.csv')\n",
    "mapping_file.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ddf25c6-2772-48a6-b4bf-aaee40e147f8",
   "metadata": {},
   "source": [
    "# Read genomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "34c97b10-86a7-4a20-9a51-d3d78802e79b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ben_mag_df = pd.read_csv('ifbdata/atlanteco_hack/MetaGenomics/BenguelaCurrent_MAGab/BenguelaCurrent_SylphResults.tsv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f1aafda7-6919-47ac-b2ed-b0923660057f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>key</th>\n",
       "      <th>ProfilingMethod</th>\n",
       "      <th>Sample-ID</th>\n",
       "      <th>Station-Label</th>\n",
       "      <th>TopicalStudy</th>\n",
       "      <th>Sample_file</th>\n",
       "      <th>Genome_file</th>\n",
       "      <th>Taxonomic_abundance</th>\n",
       "      <th>Sequence_abundance</th>\n",
       "      <th>...</th>\n",
       "      <th>Eff_lambda</th>\n",
       "      <th>Lambda_5-95_percentile</th>\n",
       "      <th>Median_cov</th>\n",
       "      <th>Mean_cov_geq1</th>\n",
       "      <th>Containment_ind</th>\n",
       "      <th>Naive_ANI</th>\n",
       "      <th>kmers_reassigned</th>\n",
       "      <th>Contig_name</th>\n",
       "      <th>GTDB_classification</th>\n",
       "      <th>NCBI_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>sylph.v0.8.0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>MMA-123</td>\n",
       "      <td>MMA-Topical-Study_Benguela-Current-Ecosystem_L...</td>\n",
       "      <td>ERR14792484_1.fastq.gz</td>\n",
       "      <td>ERR14792484_concoct_104</td>\n",
       "      <td>30.8299</td>\n",
       "      <td>2.1145</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>196.0</td>\n",
       "      <td>196.961</td>\n",
       "      <td>14025/14027</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NODE.4_length_332230_cov_108.525819</td>\n",
       "      <td>168    d__Bacteria;p__Bacteroidota;c__Bacteroi...</td>\n",
       "      <td>168    d__Bacteria;p__Bacteroidetes;c__Flavoba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>sylph.v0.8.0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>MMA-123</td>\n",
       "      <td>MMA-Topical-Study_Benguela-Current-Ecosystem_L...</td>\n",
       "      <td>ERR14792484_1.fastq.gz</td>\n",
       "      <td>ERR14792476_binner12_Refined_88</td>\n",
       "      <td>8.4939</td>\n",
       "      <td>0.6304</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>54.0</td>\n",
       "      <td>60.522</td>\n",
       "      <td>15302/15390</td>\n",
       "      <td>99.98</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NODE.6_length_154883_cov_8.411431</td>\n",
       "      <td>108    d__Bacteria;p__Pseudomonadota;c__Gammap...</td>\n",
       "      <td>108    d__Bacteria;p__Proteobacteria;c__Gammap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>sylph.v0.8.0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>MMA-123</td>\n",
       "      <td>MMA-Topical-Study_Benguela-Current-Ecosystem_L...</td>\n",
       "      <td>ERR14792484_1.fastq.gz</td>\n",
       "      <td>ERR14792494_binner12_Refined_174</td>\n",
       "      <td>7.8648</td>\n",
       "      <td>0.5930</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>50.0</td>\n",
       "      <td>47.963</td>\n",
       "      <td>15378/15594</td>\n",
       "      <td>99.96</td>\n",
       "      <td>89.0</td>\n",
       "      <td>NODE.77_length_137683_cov_7.341159</td>\n",
       "      <td>118    d__Bacteria;p__Pseudomonadota;c__Gammap...</td>\n",
       "      <td>118    d__Bacteria;p__Proteobacteria;c__Gammap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>sylph.v0.8.0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>MMA-123</td>\n",
       "      <td>MMA-Topical-Study_Benguela-Current-Ecosystem_L...</td>\n",
       "      <td>ERR14792484_1.fastq.gz</td>\n",
       "      <td>ERR14792499_binner123_Refined_14</td>\n",
       "      <td>6.2918</td>\n",
       "      <td>0.4557</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>40.0</td>\n",
       "      <td>40.616</td>\n",
       "      <td>15081/15095</td>\n",
       "      <td>100.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NODE.5_length_393469_cov_9.236133</td>\n",
       "      <td>124    d__Bacteria;p__Pseudomonadota;c__Gammap...</td>\n",
       "      <td>124    d__Bacteria;p__Proteobacteria;c__Gammap...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>sylph.v0.8.0</td>\n",
       "      <td>SAMEA7934747</td>\n",
       "      <td>MMA-123</td>\n",
       "      <td>MMA-Topical-Study_Benguela-Current-Ecosystem_L...</td>\n",
       "      <td>ERR14792484_1.fastq.gz</td>\n",
       "      <td>ERR14792491_binner13_Refined_82</td>\n",
       "      <td>5.3480</td>\n",
       "      <td>0.3840</td>\n",
       "      <td>...</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>34.0</td>\n",
       "      <td>33.151</td>\n",
       "      <td>14499/14803</td>\n",
       "      <td>99.93</td>\n",
       "      <td>21.0</td>\n",
       "      <td>NODE.196_length_68065_cov_7.101867</td>\n",
       "      <td>220    d__Bacteria;p__Pseudomonadota;c__Alphap...</td>\n",
       "      <td>220    d__Bacteria;p__Proteobacteria;c__Alphap...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0           key ProfilingMethod     Sample-ID Station-Label  \\\n",
       "0           0  SAMEA7934747    sylph.v0.8.0  SAMEA7934747       MMA-123   \n",
       "1           1  SAMEA7934747    sylph.v0.8.0  SAMEA7934747       MMA-123   \n",
       "2           2  SAMEA7934747    sylph.v0.8.0  SAMEA7934747       MMA-123   \n",
       "3           3  SAMEA7934747    sylph.v0.8.0  SAMEA7934747       MMA-123   \n",
       "4           4  SAMEA7934747    sylph.v0.8.0  SAMEA7934747       MMA-123   \n",
       "\n",
       "                                        TopicalStudy             Sample_file  \\\n",
       "0  MMA-Topical-Study_Benguela-Current-Ecosystem_L...  ERR14792484_1.fastq.gz   \n",
       "1  MMA-Topical-Study_Benguela-Current-Ecosystem_L...  ERR14792484_1.fastq.gz   \n",
       "2  MMA-Topical-Study_Benguela-Current-Ecosystem_L...  ERR14792484_1.fastq.gz   \n",
       "3  MMA-Topical-Study_Benguela-Current-Ecosystem_L...  ERR14792484_1.fastq.gz   \n",
       "4  MMA-Topical-Study_Benguela-Current-Ecosystem_L...  ERR14792484_1.fastq.gz   \n",
       "\n",
       "                        Genome_file  Taxonomic_abundance  Sequence_abundance  \\\n",
       "0           ERR14792484_concoct_104              30.8299              2.1145   \n",
       "1   ERR14792476_binner12_Refined_88               8.4939              0.6304   \n",
       "2  ERR14792494_binner12_Refined_174               7.8648              0.5930   \n",
       "3  ERR14792499_binner123_Refined_14               6.2918              0.4557   \n",
       "4   ERR14792491_binner13_Refined_82               5.3480              0.3840   \n",
       "\n",
       "   ...  Eff_lambda  Lambda_5-95_percentile Median_cov Mean_cov_geq1  \\\n",
       "0  ...        HIGH                   NA-NA      196.0       196.961   \n",
       "1  ...        HIGH                   NA-NA       54.0        60.522   \n",
       "2  ...        HIGH                   NA-NA       50.0        47.963   \n",
       "3  ...        HIGH                   NA-NA       40.0        40.616   \n",
       "4  ...        HIGH                   NA-NA       34.0        33.151   \n",
       "\n",
       "  Containment_ind  Naive_ANI  kmers_reassigned  \\\n",
       "0     14025/14027     100.00               1.0   \n",
       "1     15302/15390      99.98              26.0   \n",
       "2     15378/15594      99.96              89.0   \n",
       "3     15081/15095     100.00               2.0   \n",
       "4     14499/14803      99.93              21.0   \n",
       "\n",
       "                           Contig_name  \\\n",
       "0  NODE.4_length_332230_cov_108.525819   \n",
       "1    NODE.6_length_154883_cov_8.411431   \n",
       "2   NODE.77_length_137683_cov_7.341159   \n",
       "3    NODE.5_length_393469_cov_9.236133   \n",
       "4   NODE.196_length_68065_cov_7.101867   \n",
       "\n",
       "                                 GTDB_classification  \\\n",
       "0  168    d__Bacteria;p__Bacteroidota;c__Bacteroi...   \n",
       "1  108    d__Bacteria;p__Pseudomonadota;c__Gammap...   \n",
       "2  118    d__Bacteria;p__Pseudomonadota;c__Gammap...   \n",
       "3  124    d__Bacteria;p__Pseudomonadota;c__Gammap...   \n",
       "4  220    d__Bacteria;p__Pseudomonadota;c__Alphap...   \n",
       "\n",
       "                                 NCBI_classification  \n",
       "0  168    d__Bacteria;p__Bacteroidetes;c__Flavoba...  \n",
       "1  108    d__Bacteria;p__Proteobacteria;c__Gammap...  \n",
       "2  118    d__Bacteria;p__Proteobacteria;c__Gammap...  \n",
       "3  124    d__Bacteria;p__Proteobacteria;c__Gammap...  \n",
       "4  220    d__Bacteria;p__Proteobacteria;c__Alphap...  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ben_mag_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8682ef88-3fb2-425c-b5c1-8dfc15d597eb",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a625008-b20f-4735-871d-840673f78eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_abundances(df:pd.DataFrame, method:str='tss', id_col:str=None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Normalize gene abundance data for comparison between samples.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas.DataFrame\n",
    "        DataFrame with gene IDs in first column and samples as remaining columns\n",
    "    method : str, default 'tss'\n",
    "        Normalization method:\n",
    "        - 'tss': Total Sum Scaling (relative abundance, sums to 1)\n",
    "        - 'tss_percent': Total Sum Scaling as percentages (sums to 100)\n",
    "        - 'z_score': Z-score normalization (mean=0, std=1)\n",
    "        - 'min_max': Min-max scaling (0 to 1 range)\n",
    "        - 'log_tss': Log-transformed TSS (log10(TSS + pseudocount))\n",
    "        - 'clr': Centered Log Ratio transformation\n",
    "    id_col : str, optional\n",
    "        Name of ID column. If None, assumes first column is ID\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    pandas.DataFrame\n",
    "        Normalized DataFrame with same structure as input\n",
    "    \"\"\"\n",
    "    \n",
    "    # Make a copy to avoid modifying original\n",
    "    df_norm = df.copy()\n",
    "    \n",
    "    # Identify ID column\n",
    "    if id_col is None:\n",
    "        id_col = df.columns[0]\n",
    "    \n",
    "    # Get sample columns (all except ID column)\n",
    "    sample_cols = [col for col in df.columns if col != id_col]\n",
    "    \n",
    "    # Extract abundance matrix\n",
    "    abundance_matrix = df_norm[sample_cols].values\n",
    "    \n",
    "    if method == 'tss':\n",
    "        # Total Sum Scaling - convert to relative abundances\n",
    "        col_sums = abundance_matrix.sum(axis=0)\n",
    "        normalized_matrix = abundance_matrix / col_sums\n",
    "        \n",
    "    elif method == 'tss_percent':\n",
    "        # Total Sum Scaling as percentages\n",
    "        col_sums = abundance_matrix.sum(axis=0)\n",
    "        normalized_matrix = (abundance_matrix / col_sums) * 100\n",
    "        \n",
    "    elif method == 'z_score':\n",
    "        # Z-score normalization (standardization)\n",
    "        normalized_matrix = (abundance_matrix - abundance_matrix.mean(axis=0)) / abundance_matrix.std(axis=0)\n",
    "        \n",
    "    elif method == 'min_max':\n",
    "        # Min-max scaling to [0, 1] range\n",
    "        min_vals = abundance_matrix.min(axis=0)\n",
    "        max_vals = abundance_matrix.max(axis=0)\n",
    "        normalized_matrix = (abundance_matrix - min_vals) / (max_vals - min_vals)\n",
    "        \n",
    "    elif method == 'log_tss':\n",
    "        # Log-transformed TSS (common in metagenomics)\n",
    "        col_sums = abundance_matrix.sum(axis=0)\n",
    "        tss_matrix = abundance_matrix / col_sums\n",
    "        # Add small pseudocount to avoid log(0)\n",
    "        pseudocount = 1e-10\n",
    "        normalized_matrix = np.log10(tss_matrix + pseudocount)\n",
    "        \n",
    "    elif method == 'clr':\n",
    "        # Centered Log Ratio transformation\n",
    "        # Add small pseudocount to avoid log(0)\n",
    "        pseudocount = 1e-10\n",
    "        log_matrix = np.log(abundance_matrix + pseudocount)\n",
    "        geometric_means = log_matrix.mean(axis=0)\n",
    "        normalized_matrix = log_matrix - geometric_means\n",
    "        \n",
    "    else:\n",
    "        raise ValueError(f\"Unknown normalization method: {method}\")\n",
    "    \n",
    "    # Replace the sample columns with normalized values\n",
    "    df_norm[sample_cols] = normalized_matrix\n",
    "    \n",
    "    return df_norm\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "16432a7e-4bf4-4409-a836-cc0a29d186a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genomic_data(genomic_data_path:str, rows_to_skip:int=7) -> pd.DataFrame:\n",
    "\n",
    "    # read df\n",
    "    gen_df = pd.read_csv(genomic_data_path, sep='\\t')\n",
    "    # rename gene id row to ID\n",
    "    gen_df.rename(columns={'Unnamed: 0':SAMPLE_ID_COL}, inplace=True)\n",
    "    # remove metaparams\n",
    "    gen_df = gen_df.iloc[rows_to_skip:]\n",
    "    # convert NAN to 0\n",
    "    gen_df.fillna(0, inplace=True)\n",
    "    # change values to numeric (expect geneID)\n",
    "    sample_cols = gen_df.columns.drop(SAMPLE_ID_COL)\n",
    "    gen_df[sample_cols] = gen_df[sample_cols].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "    # Remove samples with no genes\n",
    "    # 1. Calculate column sums for sample columns\n",
    "    col_sums = gen_df[sample_cols].sum()\n",
    "    # 2. Find columns with zero sum\n",
    "    zero_sum_cols = col_sums[col_sums == 0].index.tolist()\n",
    "    # 3. Remove zero-sum columns\n",
    "    if zero_sum_cols:\n",
    "        gen_df = gen_df.drop(columns=zero_sum_cols)\n",
    "\n",
    "    # reset index\n",
    "    gen_df.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    # normalize the results per sample\n",
    "    gen_df_normalized = normalize_gene_abundances(gen_df, method='tss', id_col=SAMPLE_ID_COL)\n",
    "\n",
    "    # return normalized df\n",
    "    return gen_df_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b755b0a-6bca-4db8-950d-78619236796a",
   "metadata": {},
   "source": [
    "## Running genomic code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "0b725ddc-2185-4ec3-8ce4-5077d300c850",
   "metadata": {},
   "outputs": [],
   "source": [
    "ben_gen_df = read_genomic_data(genomic_data_path='ifbdata/atlanteco_hack/MetaGenomics/BenguelaCurrent_GeneAb/BenguelaCurrent_ffn_GeneAb_T.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "32a4a968-3490-41e3-93b3-ca4747d4c6f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample_id</th>\n",
       "      <th>ERR13506499_binner13_Refined_13_00241</th>\n",
       "      <th>RD49_S11_L002_concoct_88_00609</th>\n",
       "      <th>ERR14792500_concoct_73_00952</th>\n",
       "      <th>ERR14792466_binner12_Refined_201_02069</th>\n",
       "      <th>ERR14792484_binner12_Refined_21_01647</th>\n",
       "      <th>ERR14792474_binner12_Refined_95_02500</th>\n",
       "      <th>ERR14792503_binner123_Refined_46_00236</th>\n",
       "      <th>ERR13506423_concoct_67_00051</th>\n",
       "      <th>ERR14792468_metabat2_10_01118</th>\n",
       "      <th>ERR14792474_binner12_Refined_58_01333</th>\n",
       "      <th>...</th>\n",
       "      <th>ERR14792508_concoct_23_00832</th>\n",
       "      <th>ERR14792494_concoct_68_00603</th>\n",
       "      <th>ERR14792499_binner12_Refined_62_02272</th>\n",
       "      <th>ERR14792508_concoct_23_00544</th>\n",
       "      <th>ERR14792494_concoct_68_00409</th>\n",
       "      <th>ERR14792494_concoct_68_00496</th>\n",
       "      <th>ERR14792494_concoct_68_00014</th>\n",
       "      <th>ERR14792494_concoct_68_00186</th>\n",
       "      <th>ERR14792494_concoct_68_00822</th>\n",
       "      <th>ERR14792472_binner13_Refined_56_00115</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA7908697</th>\n",
       "      <td>0.015145</td>\n",
       "      <td>0.004184</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.889218e-06</td>\n",
       "      <td>0.000758</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>2.040104e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>9.419175e-08</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.111457e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.330850e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.131218e-06</td>\n",
       "      <td>2.427111e-07</td>\n",
       "      <td>0.000220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7909194</th>\n",
       "      <td>0.033343</td>\n",
       "      <td>0.009253</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.874588e-07</td>\n",
       "      <td>0.007311</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>5.710084e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.323540e-05</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>7.543917e-06</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>4.818135e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.643667e-05</td>\n",
       "      <td>7.650181e-07</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7913651</th>\n",
       "      <td>0.794466</td>\n",
       "      <td>0.198864</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7926718</th>\n",
       "      <td>0.072409</td>\n",
       "      <td>0.020711</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.016018e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.119132e-07</td>\n",
       "      <td>0.018887</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>6.549923e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.712454e-05</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>6.050491e-06</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>7.360560e-05</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>6.763142e-05</td>\n",
       "      <td>1.782212e-05</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7896668</th>\n",
       "      <td>0.013803</td>\n",
       "      <td>0.003844</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>1.346209e-06</td>\n",
       "      <td>0.001491</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>8.691979e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.579185e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.408541e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000269</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample_id     ERR13506499_binner13_Refined_13_00241  \\\n",
       "SAMEA7908697                               0.015145   \n",
       "SAMEA7909194                               0.033343   \n",
       "SAMEA7913651                               0.794466   \n",
       "SAMEA7926718                               0.072409   \n",
       "SAMEA7896668                               0.013803   \n",
       "\n",
       "sample_id     RD49_S11_L002_concoct_88_00609  ERR14792500_concoct_73_00952  \\\n",
       "SAMEA7908697                        0.004184                      0.000002   \n",
       "SAMEA7909194                        0.009253                      0.000000   \n",
       "SAMEA7913651                        0.198864                      0.000000   \n",
       "SAMEA7926718                        0.020711                      0.000000   \n",
       "SAMEA7896668                        0.003844                      0.000000   \n",
       "\n",
       "sample_id     ERR14792466_binner12_Refined_201_02069  \\\n",
       "SAMEA7908697                                0.000000   \n",
       "SAMEA7909194                                0.000000   \n",
       "SAMEA7913651                                0.000000   \n",
       "SAMEA7926718                                0.000000   \n",
       "SAMEA7896668                                0.000003   \n",
       "\n",
       "sample_id     ERR14792484_binner12_Refined_21_01647  \\\n",
       "SAMEA7908697                           0.000000e+00   \n",
       "SAMEA7909194                           0.000000e+00   \n",
       "SAMEA7913651                           0.000000e+00   \n",
       "SAMEA7926718                           3.016018e-07   \n",
       "SAMEA7896668                           0.000000e+00   \n",
       "\n",
       "sample_id     ERR14792474_binner12_Refined_95_02500  \\\n",
       "SAMEA7908697                               0.000000   \n",
       "SAMEA7909194                               0.000000   \n",
       "SAMEA7913651                               0.000000   \n",
       "SAMEA7926718                               0.000000   \n",
       "SAMEA7896668                               0.000004   \n",
       "\n",
       "sample_id     ERR14792503_binner123_Refined_46_00236  \\\n",
       "SAMEA7908697                            3.889218e-06   \n",
       "SAMEA7909194                            4.874588e-07   \n",
       "SAMEA7913651                            0.000000e+00   \n",
       "SAMEA7926718                            5.119132e-07   \n",
       "SAMEA7896668                            1.346209e-06   \n",
       "\n",
       "sample_id     ERR13506423_concoct_67_00051  ERR14792468_metabat2_10_01118  \\\n",
       "SAMEA7908697                      0.000758                       0.000016   \n",
       "SAMEA7909194                      0.007311                       0.000003   \n",
       "SAMEA7913651                      0.000000                       0.000000   \n",
       "SAMEA7926718                      0.018887                       0.000002   \n",
       "SAMEA7896668                      0.001491                       0.000005   \n",
       "\n",
       "sample_id     ERR14792474_binner12_Refined_58_01333  ...  \\\n",
       "SAMEA7908697                           2.040104e-06  ...   \n",
       "SAMEA7909194                           5.710084e-07  ...   \n",
       "SAMEA7913651                           0.000000e+00  ...   \n",
       "SAMEA7926718                           6.549923e-07  ...   \n",
       "SAMEA7896668                           8.691979e-06  ...   \n",
       "\n",
       "sample_id     ERR14792508_concoct_23_00832  ERR14792494_concoct_68_00603  \\\n",
       "SAMEA7908697                  9.419175e-08                      0.000000   \n",
       "SAMEA7909194                  1.323540e-05                      0.000002   \n",
       "SAMEA7913651                  0.000000e+00                      0.000000   \n",
       "SAMEA7926718                  4.712454e-05                      0.000009   \n",
       "SAMEA7896668                  0.000000e+00                      0.000000   \n",
       "\n",
       "sample_id     ERR14792499_binner12_Refined_62_02272  \\\n",
       "SAMEA7908697                           1.111457e-05   \n",
       "SAMEA7909194                           7.543917e-06   \n",
       "SAMEA7913651                           0.000000e+00   \n",
       "SAMEA7926718                           6.050491e-06   \n",
       "SAMEA7896668                           5.579185e-07   \n",
       "\n",
       "sample_id     ERR14792508_concoct_23_00544  ERR14792494_concoct_68_00409  \\\n",
       "SAMEA7908697                      0.000000                      0.000000   \n",
       "SAMEA7909194                      0.000003                      0.000014   \n",
       "SAMEA7913651                      0.000000                      0.000000   \n",
       "SAMEA7926718                      0.000015                      0.000118   \n",
       "SAMEA7896668                      0.000000                      0.000000   \n",
       "\n",
       "sample_id     ERR14792494_concoct_68_00496  ERR14792494_concoct_68_00014  \\\n",
       "SAMEA7908697                  4.330850e-07                      0.000000   \n",
       "SAMEA7909194                  4.818135e-06                      0.000000   \n",
       "SAMEA7913651                  0.000000e+00                      0.000000   \n",
       "SAMEA7926718                  7.360560e-05                      0.000094   \n",
       "SAMEA7896668                  0.000000e+00                      0.000000   \n",
       "\n",
       "sample_id     ERR14792494_concoct_68_00186  ERR14792494_concoct_68_00822  \\\n",
       "SAMEA7908697                  1.131218e-06                  2.427111e-07   \n",
       "SAMEA7909194                  1.643667e-05                  7.650181e-07   \n",
       "SAMEA7913651                  0.000000e+00                  0.000000e+00   \n",
       "SAMEA7926718                  6.763142e-05                  1.782212e-05   \n",
       "SAMEA7896668                  2.408541e-08                  0.000000e+00   \n",
       "\n",
       "sample_id     ERR14792472_binner13_Refined_56_00115  \n",
       "SAMEA7908697                               0.000220  \n",
       "SAMEA7909194                               0.000033  \n",
       "SAMEA7913651                               0.000000  \n",
       "SAMEA7926718                               0.000003  \n",
       "SAMEA7896668                               0.000269  \n",
       "\n",
       "[5 rows x 200 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_200_ben_genes = get_most_variable_omics(ben_gen_df, 200)\n",
    "top_200_ben_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c697af46-ab65-42c0-b525-835f9d546670",
   "metadata": {},
   "outputs": [],
   "source": [
    "# wedd_gen_df = read_genomic_data(genomic_data_path='ifbdata/atlanteco_hack/MetaGenomics/WeddellSea_GeneAb/WeddellSea_ffn_GeneAb_T.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cebd9da4-d6f2-4be7-b290-7d82294068bf",
   "metadata": {},
   "source": [
    "# Read environmental data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be7a687-603d-4c7c-9c60-6133f1f5ce71",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "972dcbea-6018-41ee-92fe-a47b2ee486c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_env_data(env_data_csv_path:str) -> pd.DataFrame:\n",
    "    return pd.read_csv(env_data_csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9a5320e4-1711-4e25-bed5-2516e655e303",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>MMA-071</th>\n",
       "      <th>MMA-072</th>\n",
       "      <th>MMA-074</th>\n",
       "      <th>MMA-076</th>\n",
       "      <th>MMA-077</th>\n",
       "      <th>MMA-078</th>\n",
       "      <th>MMA-079</th>\n",
       "      <th>MMA-080</th>\n",
       "      <th>MMA-081</th>\n",
       "      <th>...</th>\n",
       "      <th>MMA-118</th>\n",
       "      <th>MMA-120</th>\n",
       "      <th>MMA-121</th>\n",
       "      <th>MMA-122</th>\n",
       "      <th>MMA-123</th>\n",
       "      <th>MMA-124</th>\n",
       "      <th>MMA-127</th>\n",
       "      <th>MMA-130</th>\n",
       "      <th>MMA-131</th>\n",
       "      <th>MMA-132</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTD temperature (degC) median</td>\n",
       "      <td>-0.286</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-1.0150</td>\n",
       "      <td>-0.8130</td>\n",
       "      <td>-0.9110</td>\n",
       "      <td>-0.9130</td>\n",
       "      <td>-0.943</td>\n",
       "      <td>-1.0330</td>\n",
       "      <td>-0.922</td>\n",
       "      <td>...</td>\n",
       "      <td>15.2730</td>\n",
       "      <td>17.1460</td>\n",
       "      <td>15.1610</td>\n",
       "      <td>16.6450</td>\n",
       "      <td>16.8800</td>\n",
       "      <td>13.9100</td>\n",
       "      <td>20.9720</td>\n",
       "      <td>24.0940</td>\n",
       "      <td>23.7460</td>\n",
       "      <td>24.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CTD salinity (psu) median</td>\n",
       "      <td>34.422</td>\n",
       "      <td>33.9444</td>\n",
       "      <td>33.0256</td>\n",
       "      <td>33.0732</td>\n",
       "      <td>33.1779</td>\n",
       "      <td>33.1789</td>\n",
       "      <td>33.167</td>\n",
       "      <td>33.2374</td>\n",
       "      <td>33.360</td>\n",
       "      <td>...</td>\n",
       "      <td>35.1678</td>\n",
       "      <td>35.5247</td>\n",
       "      <td>35.4309</td>\n",
       "      <td>35.3534</td>\n",
       "      <td>35.4706</td>\n",
       "      <td>35.3161</td>\n",
       "      <td>36.2856</td>\n",
       "      <td>29.0101</td>\n",
       "      <td>30.8759</td>\n",
       "      <td>36.0856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTD oxygen (%) median</td>\n",
       "      <td>93.000</td>\n",
       "      <td>104.8000</td>\n",
       "      <td>95.2000</td>\n",
       "      <td>95.3000</td>\n",
       "      <td>94.6000</td>\n",
       "      <td>94.5000</td>\n",
       "      <td>94.600</td>\n",
       "      <td>93.9000</td>\n",
       "      <td>94.300</td>\n",
       "      <td>...</td>\n",
       "      <td>99.3000</td>\n",
       "      <td>89.3000</td>\n",
       "      <td>63.2000</td>\n",
       "      <td>97.1000</td>\n",
       "      <td>92.7000</td>\n",
       "      <td>34.7000</td>\n",
       "      <td>104.4000</td>\n",
       "      <td>79.8000</td>\n",
       "      <td>76.1000</td>\n",
       "      <td>80.7000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTD fluorescence (mg/m3) median</td>\n",
       "      <td>123.569</td>\n",
       "      <td>48.7230</td>\n",
       "      <td>4.3520</td>\n",
       "      <td>1.8870</td>\n",
       "      <td>3.7230</td>\n",
       "      <td>2.8450</td>\n",
       "      <td>2.085</td>\n",
       "      <td>1.8150</td>\n",
       "      <td>2.734</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0850</td>\n",
       "      <td>4.0080</td>\n",
       "      <td>4.3500</td>\n",
       "      <td>8.7580</td>\n",
       "      <td>5.6290</td>\n",
       "      <td>5.8940</td>\n",
       "      <td>5.1880</td>\n",
       "      <td>1.5250</td>\n",
       "      <td>1.8440</td>\n",
       "      <td>2.0290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NO2</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.0800</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>0.2200</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>0.2100</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.1700</td>\n",
       "      <td>0.170</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4600</td>\n",
       "      <td>0.5000</td>\n",
       "      <td>0.5800</td>\n",
       "      <td>0.1700</td>\n",
       "      <td>0.3700</td>\n",
       "      <td>0.6200</td>\n",
       "      <td>0.1600</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature  MMA-071   MMA-072  MMA-074  MMA-076  \\\n",
       "0    CTD temperature (degC) median   -0.286   -0.1140  -1.0150  -0.8130   \n",
       "1        CTD salinity (psu) median   34.422   33.9444  33.0256  33.0732   \n",
       "2            CTD oxygen (%) median   93.000  104.8000  95.2000  95.3000   \n",
       "3  CTD fluorescence (mg/m3) median  123.569   48.7230   4.3520   1.8870   \n",
       "4                              NO2    0.150    0.0800   0.1900   0.2200   \n",
       "\n",
       "   MMA-077  MMA-078  MMA-079  MMA-080  MMA-081  ...  MMA-118  MMA-120  \\\n",
       "0  -0.9110  -0.9130   -0.943  -1.0330   -0.922  ...  15.2730  17.1460   \n",
       "1  33.1779  33.1789   33.167  33.2374   33.360  ...  35.1678  35.5247   \n",
       "2  94.6000  94.5000   94.600  93.9000   94.300  ...  99.3000  89.3000   \n",
       "3   3.7230   2.8450    2.085   1.8150    2.734  ...   9.0850   4.0080   \n",
       "4   0.1900   0.2100    0.200   0.1700    0.170  ...   0.4600   0.5000   \n",
       "\n",
       "   MMA-121  MMA-122  MMA-123  MMA-124   MMA-127  MMA-130  MMA-131  MMA-132  \n",
       "0  15.1610  16.6450  16.8800  13.9100   20.9720  24.0940  23.7460  24.2430  \n",
       "1  35.4309  35.3534  35.4706  35.3161   36.2856  29.0101  30.8759  36.0856  \n",
       "2  63.2000  97.1000  92.7000  34.7000  104.4000  79.8000  76.1000  80.7000  \n",
       "3   4.3500   8.7580   5.6290   5.8940    5.1880   1.5250   1.8440   2.0290  \n",
       "4   0.5800   0.1700   0.3700   0.6200    0.1600   0.0000   0.0000   0.0100  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_df = read_env_data('machine_ready_env.csv')\n",
    "env_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1d50aa9-0694-4414-86ad-28e312342858",
   "metadata": {},
   "source": [
    "# Read Proteomics data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7ced78b-8ac7-4ccb-bcf7-7afbddcb0dbb",
   "metadata": {},
   "source": [
    "## Helper function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "fb8064fe-e21b-41f6-8c11-124e224ffbd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_prot_data(prot_data_path:str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(prot_data_path)\n",
    "    df.columns = df.columns.str.replace('_normalized', '')\n",
    "    df.rename(columns={\"MAG_name_prot\":SAMPLE_ID_COL}, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767128ed-cc1e-4a33-be30-7be742fff209",
   "metadata": {},
   "source": [
    "## Running proteomic code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "c9d4c301-c0af-462a-b8de-8a3a78456b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "ben_prot_df = read_prot_data('annotation_normalized_benguela.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "79043906-332d-4738-9e1e-88c3a659c476",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample_id</th>\n",
       "      <th>megahit_HN020_k141_129271.p1</th>\n",
       "      <th>transabyss_HN031_.k50S2815693.p1</th>\n",
       "      <th>transabyss_HN017_.k50R6170578.p2</th>\n",
       "      <th>transabyss_HN006_.k50R5032280.p1</th>\n",
       "      <th>trinity_HN021_TRINITY_DN12475_c0_g1_i1.p3</th>\n",
       "      <th>megahit_HN002_k141_333561.p1</th>\n",
       "      <th>transabyss_HN026_.k50R3537400.p2</th>\n",
       "      <th>trinity_HN039_TRINITY_DN8611_c0_g1_i9.p1</th>\n",
       "      <th>megahit_HN002_k141_173328.p1</th>\n",
       "      <th>megahit_HN011_k141_216036.p1</th>\n",
       "      <th>...</th>\n",
       "      <th>trinity_HN002_TRINITY_DN12831_c0_g2_i2.p1</th>\n",
       "      <th>megahit_HN022_k141_181984.p1</th>\n",
       "      <th>megahit_HN016_k141_237681.p1</th>\n",
       "      <th>megahit_HN016_k141_148937.p1</th>\n",
       "      <th>transabyss_HN042_.k20R32033753.p1</th>\n",
       "      <th>trinity_HN011_TRINITY_DN58831_c0_g1_i1.p1</th>\n",
       "      <th>megahit_HN006_k141_164784.p1</th>\n",
       "      <th>megahit_HN011_k141_520560.p2</th>\n",
       "      <th>trinity_HN017_TRINITY_DN29133_c0_g1_i3.p2</th>\n",
       "      <th>trinity_HN038_TRINITY_DN19851_c0_g1_i1.p1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMEA7917467</th>\n",
       "      <td>1.380000e-07</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>5.550000e-07</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>9.430000e-07</td>\n",
       "      <td>2.610000e-06</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.570000e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.300000e-07</td>\n",
       "      <td>1.980000e-06</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>6.190000e-07</td>\n",
       "      <td>8.150000e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.680000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.660000e-06</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7917491</th>\n",
       "      <td>8.690000e-07</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>3.490000e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>3.920000e-06</td>\n",
       "      <td>4.680000e-07</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>3.670000e-07</td>\n",
       "      <td>4.540000e-06</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>5.890000e-07</td>\n",
       "      <td>1.260000e-06</td>\n",
       "      <td>6.950000e-08</td>\n",
       "      <td>2.740000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.450000e-06</td>\n",
       "      <td>0.000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7917515</th>\n",
       "      <td>1.360000e-07</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>4.960000e-06</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>1.070000e-06</td>\n",
       "      <td>1.290000e-06</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>9.560000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.150000e-06</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>3.970000e-07</td>\n",
       "      <td>1.420000e-06</td>\n",
       "      <td>1.740000e-07</td>\n",
       "      <td>7.100000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.010000e-06</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7908692</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>5.360000e-06</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>5.780000e-07</td>\n",
       "      <td>2.330000e-06</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>1.060000e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>8.990000e-08</td>\n",
       "      <td>2.370000e-06</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>7.080000e-07</td>\n",
       "      <td>1.550000e-06</td>\n",
       "      <td>2.340000e-08</td>\n",
       "      <td>3.070000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.300000e-07</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA7908700</th>\n",
       "      <td>6.560000e-08</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>1.540000e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.260000e-06</td>\n",
       "      <td>4.560000e-06</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>2.990000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.080000e-07</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>8.830000e-08</td>\n",
       "      <td>4.670000e-07</td>\n",
       "      <td>2.850000e-08</td>\n",
       "      <td>1.300000e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.790000e-07</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample_id     megahit_HN020_k141_129271.p1  transabyss_HN031_.k50S2815693.p1  \\\n",
       "SAMEA7917467                  1.380000e-07                          0.000038   \n",
       "SAMEA7917491                  8.690000e-07                          0.000032   \n",
       "SAMEA7917515                  1.360000e-07                          0.000037   \n",
       "SAMEA7908692                  0.000000e+00                          0.000034   \n",
       "SAMEA7908700                  6.560000e-08                          0.000039   \n",
       "\n",
       "sample_id     transabyss_HN017_.k50R6170578.p2  \\\n",
       "SAMEA7917467                      5.550000e-07   \n",
       "SAMEA7917491                      3.490000e-07   \n",
       "SAMEA7917515                      4.960000e-06   \n",
       "SAMEA7908692                      5.360000e-06   \n",
       "SAMEA7908700                      1.540000e-06   \n",
       "\n",
       "sample_id     transabyss_HN006_.k50R5032280.p1  \\\n",
       "SAMEA7917467                          0.000002   \n",
       "SAMEA7917491                          0.000000   \n",
       "SAMEA7917515                          0.000004   \n",
       "SAMEA7908692                          0.000002   \n",
       "SAMEA7908700                          0.000000   \n",
       "\n",
       "sample_id     trinity_HN021_TRINITY_DN12475_c0_g1_i1.p3  \\\n",
       "SAMEA7917467                                   0.000004   \n",
       "SAMEA7917491                                   0.000002   \n",
       "SAMEA7917515                                   0.000004   \n",
       "SAMEA7908692                                   0.000018   \n",
       "SAMEA7908700                                   0.000002   \n",
       "\n",
       "sample_id     megahit_HN002_k141_333561.p1  transabyss_HN026_.k50R3537400.p2  \\\n",
       "SAMEA7917467                      0.000005                      9.430000e-07   \n",
       "SAMEA7917491                      0.000007                      3.920000e-06   \n",
       "SAMEA7917515                      0.000003                      1.070000e-06   \n",
       "SAMEA7908692                      0.000008                      5.780000e-07   \n",
       "SAMEA7908700                      0.000000                      1.260000e-06   \n",
       "\n",
       "sample_id     trinity_HN039_TRINITY_DN8611_c0_g1_i9.p1  \\\n",
       "SAMEA7917467                              2.610000e-06   \n",
       "SAMEA7917491                              4.680000e-07   \n",
       "SAMEA7917515                              1.290000e-06   \n",
       "SAMEA7908692                              2.330000e-06   \n",
       "SAMEA7908700                              4.560000e-06   \n",
       "\n",
       "sample_id     megahit_HN002_k141_173328.p1  megahit_HN011_k141_216036.p1  ...  \\\n",
       "SAMEA7917467                      0.000019                  1.570000e-07  ...   \n",
       "SAMEA7917491                      0.000043                  0.000000e+00  ...   \n",
       "SAMEA7917515                      0.000029                  9.560000e-08  ...   \n",
       "SAMEA7908692                      0.000030                  1.060000e-07  ...   \n",
       "SAMEA7908700                      0.000015                  2.990000e-08  ...   \n",
       "\n",
       "sample_id     trinity_HN002_TRINITY_DN12831_c0_g2_i2.p1  \\\n",
       "SAMEA7917467                               1.300000e-07   \n",
       "SAMEA7917491                               3.670000e-07   \n",
       "SAMEA7917515                               0.000000e+00   \n",
       "SAMEA7908692                               8.990000e-08   \n",
       "SAMEA7908700                               0.000000e+00   \n",
       "\n",
       "sample_id     megahit_HN022_k141_181984.p1  megahit_HN016_k141_237681.p1  \\\n",
       "SAMEA7917467                  1.980000e-06                      0.000001   \n",
       "SAMEA7917491                  4.540000e-06                      0.000003   \n",
       "SAMEA7917515                  1.150000e-06                      0.000002   \n",
       "SAMEA7908692                  2.370000e-06                      0.000004   \n",
       "SAMEA7908700                  1.080000e-07                      0.000001   \n",
       "\n",
       "sample_id     megahit_HN016_k141_148937.p1  transabyss_HN042_.k20R32033753.p1  \\\n",
       "SAMEA7917467                  6.190000e-07                       8.150000e-07   \n",
       "SAMEA7917491                  5.890000e-07                       1.260000e-06   \n",
       "SAMEA7917515                  3.970000e-07                       1.420000e-06   \n",
       "SAMEA7908692                  7.080000e-07                       1.550000e-06   \n",
       "SAMEA7908700                  8.830000e-08                       4.670000e-07   \n",
       "\n",
       "sample_id     trinity_HN011_TRINITY_DN58831_c0_g1_i1.p1  \\\n",
       "SAMEA7917467                               0.000000e+00   \n",
       "SAMEA7917491                               6.950000e-08   \n",
       "SAMEA7917515                               1.740000e-07   \n",
       "SAMEA7908692                               2.340000e-08   \n",
       "SAMEA7908700                               2.850000e-08   \n",
       "\n",
       "sample_id     megahit_HN006_k141_164784.p1  megahit_HN011_k141_520560.p2  \\\n",
       "SAMEA7917467                  3.680000e-07                           0.0   \n",
       "SAMEA7917491                  2.740000e-07                           0.0   \n",
       "SAMEA7917515                  7.100000e-07                           0.0   \n",
       "SAMEA7908692                  3.070000e-07                           0.0   \n",
       "SAMEA7908700                  1.300000e-07                           0.0   \n",
       "\n",
       "sample_id     trinity_HN017_TRINITY_DN29133_c0_g1_i3.p2  \\\n",
       "SAMEA7917467                               3.660000e-06   \n",
       "SAMEA7917491                               6.450000e-06   \n",
       "SAMEA7917515                               5.010000e-06   \n",
       "SAMEA7908692                               4.300000e-07   \n",
       "SAMEA7908700                               1.790000e-07   \n",
       "\n",
       "sample_id     trinity_HN038_TRINITY_DN19851_c0_g1_i1.p1  \n",
       "SAMEA7917467                                   0.000002  \n",
       "SAMEA7917491                                   0.000004  \n",
       "SAMEA7917515                                   0.000003  \n",
       "SAMEA7908692                                   0.000003  \n",
       "SAMEA7908700                                   0.000002  \n",
       "\n",
       "[5 rows x 200 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_200_ben_prots = get_most_variable_omics(ben_prot_df, 200)\n",
    "top_200_ben_prots.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "44f4a5cc-f9bb-4a56-b3c9-539113b539f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MAG_name_prot</th>\n",
       "      <th>SAMEA7917467</th>\n",
       "      <th>SAMEA7917491</th>\n",
       "      <th>SAMEA7917515</th>\n",
       "      <th>SAMEA7908692</th>\n",
       "      <th>SAMEA7908700</th>\n",
       "      <th>SAMEA7909095</th>\n",
       "      <th>SAMEA7909203</th>\n",
       "      <th>SAMEA7909267</th>\n",
       "      <th>SAMEA7913658</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMEA7934767</th>\n",
       "      <th>SAMEA7934820</th>\n",
       "      <th>SAMEA7934998</th>\n",
       "      <th>SAMEA7914376</th>\n",
       "      <th>SAMEA7935115</th>\n",
       "      <th>SAMEA7939692</th>\n",
       "      <th>SAMEA7926962</th>\n",
       "      <th>SAMEA7939770</th>\n",
       "      <th>SAMEA7927003</th>\n",
       "      <th>SAMEA7927059</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>megahit_HN020_k141_129271.p1</td>\n",
       "      <td>1.380000e-07</td>\n",
       "      <td>8.690000e-07</td>\n",
       "      <td>1.360000e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>6.560000e-08</td>\n",
       "      <td>4.430000e-08</td>\n",
       "      <td>7.800000e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.280000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.470000e-07</td>\n",
       "      <td>1.830000e-08</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>3.880000e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.230000e-07</td>\n",
       "      <td>8.940000e-08</td>\n",
       "      <td>0.000332</td>\n",
       "      <td>3.690000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trinity_HN006_TRINITY_DN216338_c0_g1_i1.p1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>megahit_HN025_k141_9493.p1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>transabyss_HN006_.k50R5032280.p1</td>\n",
       "      <td>2.390000e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.810000e-06</td>\n",
       "      <td>2.200000e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.880000e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.940000e-06</td>\n",
       "      <td>3.380000e-06</td>\n",
       "      <td>1.270000e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.750000e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.960000e-06</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>5.680000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transabyss_HN026_.k50R3537400.p2</td>\n",
       "      <td>9.430000e-07</td>\n",
       "      <td>3.920000e-06</td>\n",
       "      <td>1.070000e-06</td>\n",
       "      <td>5.780000e-07</td>\n",
       "      <td>1.260000e-06</td>\n",
       "      <td>6.910000e-08</td>\n",
       "      <td>3.480000e-07</td>\n",
       "      <td>7.590000e-07</td>\n",
       "      <td>2.940000e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.950000e-07</td>\n",
       "      <td>8.660000e-07</td>\n",
       "      <td>1.780000e-07</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>5.070000e-07</td>\n",
       "      <td>2.540000e-07</td>\n",
       "      <td>1.200000e-06</td>\n",
       "      <td>2.190000e-06</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>1.580000e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                MAG_name_prot  SAMEA7917467  SAMEA7917491  \\\n",
       "0                megahit_HN020_k141_129271.p1  1.380000e-07  8.690000e-07   \n",
       "1  trinity_HN006_TRINITY_DN216338_c0_g1_i1.p1  0.000000e+00  0.000000e+00   \n",
       "2                  megahit_HN025_k141_9493.p1  0.000000e+00  0.000000e+00   \n",
       "3            transabyss_HN006_.k50R5032280.p1  2.390000e-06  0.000000e+00   \n",
       "4            transabyss_HN026_.k50R3537400.p2  9.430000e-07  3.920000e-06   \n",
       "\n",
       "   SAMEA7917515  SAMEA7908692  SAMEA7908700  SAMEA7909095  SAMEA7909203  \\\n",
       "0  1.360000e-07  0.000000e+00  6.560000e-08  4.430000e-08  7.800000e-08   \n",
       "1  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "2  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "3  3.810000e-06  2.200000e-06  0.000000e+00  2.880000e-07  0.000000e+00   \n",
       "4  1.070000e-06  5.780000e-07  1.260000e-06  6.910000e-08  3.480000e-07   \n",
       "\n",
       "   SAMEA7909267  SAMEA7913658  ...  SAMEA7934767  SAMEA7934820  SAMEA7934998  \\\n",
       "0  0.000000e+00  3.280000e-08  ...  0.000000e+00  1.470000e-07  1.830000e-08   \n",
       "1  0.000000e+00  0.000000e+00  ...  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "2  0.000000e+00  0.000000e+00  ...  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "3  0.000000e+00  0.000000e+00  ...  1.940000e-06  3.380000e-06  1.270000e-06   \n",
       "4  7.590000e-07  2.940000e-07  ...  3.950000e-07  8.660000e-07  1.780000e-07   \n",
       "\n",
       "   SAMEA7914376  SAMEA7935115  SAMEA7939692  SAMEA7926962  SAMEA7939770  \\\n",
       "0      0.000002  3.880000e-08  0.000000e+00  1.230000e-07  8.940000e-08   \n",
       "1      0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "2      0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   \n",
       "3      0.000000  0.000000e+00  5.750000e-07  0.000000e+00  4.960000e-06   \n",
       "4      0.000001  5.070000e-07  2.540000e-07  1.200000e-06  2.190000e-06   \n",
       "\n",
       "   SAMEA7927003  SAMEA7927059  \n",
       "0      0.000332  3.690000e-07  \n",
       "1      0.000000  0.000000e+00  \n",
       "2      0.000000  0.000000e+00  \n",
       "3      0.000150  5.680000e-06  \n",
       "4      0.000119  1.580000e-06  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ben_prot_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56059c00-3993-4d73-8267-96f76c5f93fe",
   "metadata": {},
   "source": [
    "# Read Metabolomic data (labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e222e75c-8192-41de-9e48-f49437a3a62b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7475e099-e649-46d3-9267-5d269d0edd25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_relevant_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Keep only columns starting with 'featureId' or 'SAMEA'.\n",
    "    \"\"\"\n",
    "    return df.loc[:, df.columns.str.startswith(('featureId', 'SAMEA'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cc80d65-1f8f-433c-82e5-52ca87602535",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_samea_column_names(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Remove the trailing '_RX' from SAMEA column names.\n",
    "    E.g., SAMEA123456_R01_R2 -> SAMEA123456_R01\n",
    "    \"\"\"\n",
    "    rename_map = {\n",
    "        col: re.sub(r'(SAMEA\\d+_R\\d+)_R\\d+$', r'\\1', col)\n",
    "        if col.startswith('SAMEA') else col\n",
    "        for col in df.columns\n",
    "    }\n",
    "    return df.rename(columns=rename_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "19dece8a-8641-454c-8a43-5f89fed618ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_samea_columns(df: pd.DataFrame, threshold: float = 2e4) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Group columns with the same SAMEAXXXXXX prefix:\n",
    "      - If any column in the group > threshold → grouped value = 1\n",
    "      - If all columns in the group ≤ threshold → grouped value = 0\n",
    "\n",
    "    The grouped columns will replace the original SAMEA columns.\n",
    "    \"\"\"\n",
    "    # Map each SAMEA column to its base prefix (SAMEAXXXXXX)\n",
    "    prefix_map = {\n",
    "        col: re.match(r'(SAMEA\\d+)', col).group(1)\n",
    "        if col.startswith('SAMEA') else col\n",
    "        for col in df.columns\n",
    "    }\n",
    "\n",
    "    result = df.copy()\n",
    "    for prefix in set(prefix_map.values()):\n",
    "        if prefix.startswith('SAMEA'):\n",
    "            same_cols = [col for col, pfx in prefix_map.items() if pfx == prefix]\n",
    "\n",
    "            # 🔹 Ensure these columns are numeric (convert strings → numbers, non-numeric → NaN)\n",
    "            numeric_block = df[same_cols].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "            # ✅ Compute on numeric_block, not df\n",
    "            result[prefix] = (numeric_block > threshold).any(axis=1).astype(int)\n",
    "\n",
    "            # Drop original group columns\n",
    "            result = result.drop(columns=same_cols)\n",
    "\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0511265-6b33-483a-832f-da91bd8df6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_metabolomic_data(metabolome_data_path:str, threshold: float = 2e4) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Full pipeline:\n",
    "      1. Filter columns\n",
    "      2. Clean SAMEA column names\n",
    "      3. Group SAMEA columns using threshold logic\n",
    "    \"\"\"\n",
    "    metabolome_df = pd.read_csv(metabolome_data_path, sep='\\t')\n",
    "    metabolome_df.drop(metabolome_df.tail(1).index,inplace=True) # drop last row\n",
    "    metabolome_df = filter_relevant_columns(metabolome_df)\n",
    "    metabolome_df = clean_samea_column_names(metabolome_df)\n",
    "    metabolome_df = group_samea_columns(metabolome_df, threshold=threshold)\n",
    "    return metabolome_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5f5c02-77d5-42b0-a0a6-264f7ee0eed2",
   "metadata": {},
   "source": [
    "## Running metabolomic code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2d74ca0a-38e9-4c71-9cd4-d3510e4966a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolome_path = 'ifbdata/atlanteco_hack/MetaMetabolomics/1_Feature_table_univariate_analysis_hackathon'\n",
    "\n",
    "processed_metabolomic_data = process_metabolomic_data(metabolome_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e770466a-4c7e-46ca-b11a-7e4db05aab65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1    651561\n",
      "0     44715\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "sample_cols = [col for col in processed_metabolomic_data.columns if col != 'featureId']\n",
    "counts = processed_metabolomic_data[sample_cols].stack().value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4d061fdb-447e-4895-9c32-7f7bdb0a1938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1842"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(processed_metabolomic_data['featureId'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "51196945-38f7-4bf7-9849-b45970583e21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureId</th>\n",
       "      <th>SAMEA7951578</th>\n",
       "      <th>SAMEA7940389</th>\n",
       "      <th>SAMEA7920182</th>\n",
       "      <th>SAMEA7920184</th>\n",
       "      <th>SAMEA7928500</th>\n",
       "      <th>SAMEA7908666</th>\n",
       "      <th>SAMEA7955281</th>\n",
       "      <th>SAMEA7948843</th>\n",
       "      <th>SAMEA7904744</th>\n",
       "      <th>...</th>\n",
       "      <th>SAMEA7908718</th>\n",
       "      <th>SAMEA7935443</th>\n",
       "      <th>SAMEA7928718</th>\n",
       "      <th>SAMEA7950917</th>\n",
       "      <th>SAMEA7951482</th>\n",
       "      <th>SAMEA7904605</th>\n",
       "      <th>SAMEA7928635</th>\n",
       "      <th>SAMEA7909128</th>\n",
       "      <th>SAMEA7954424</th>\n",
       "      <th>SAMEA7927092</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FT00002</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FT00005</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FT00017</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FT00112</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FT00326</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 379 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  featureId  SAMEA7951578  SAMEA7940389  SAMEA7920182  SAMEA7920184  \\\n",
       "0   FT00002             1             1             1             1   \n",
       "1   FT00005             1             1             1             1   \n",
       "2   FT00017             1             1             1             1   \n",
       "3   FT00112             1             1             1             1   \n",
       "4   FT00326             1             1             1             1   \n",
       "\n",
       "   SAMEA7928500  SAMEA7908666  SAMEA7955281  SAMEA7948843  SAMEA7904744  ...  \\\n",
       "0             1             1             1             1             1  ...   \n",
       "1             1             1             0             1             1  ...   \n",
       "2             1             1             1             1             1  ...   \n",
       "3             1             1             1             1             1  ...   \n",
       "4             1             1             1             1             1  ...   \n",
       "\n",
       "   SAMEA7908718  SAMEA7935443  SAMEA7928718  SAMEA7950917  SAMEA7951482  \\\n",
       "0             1             1             1             1             1   \n",
       "1             1             1             1             1             1   \n",
       "2             1             1             1             1             1   \n",
       "3             1             1             1             1             1   \n",
       "4             1             1             1             1             1   \n",
       "\n",
       "   SAMEA7904605  SAMEA7928635  SAMEA7909128  SAMEA7954424  SAMEA7927092  \n",
       "0             1             1             1             1             1  \n",
       "1             1             1             1             1             1  \n",
       "2             1             1             1             1             1  \n",
       "3             1             1             1             1             1  \n",
       "4             1             1             1             1             1  \n",
       "\n",
       "[5 rows x 379 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_metabolomic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73b5de33-3050-45df-affa-6d6ec6b27382",
   "metadata": {},
   "source": [
    "# Mapping samples to sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "0bd1edc7-50d6-4611-8ff5-30031303934e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = dict(zip(mapping_file['sample_id'], mapping_file['sampling_design_feature']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "48eb7f2a-c5b4-4f46-90f4-d2cb6dfea45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_200_ben_prots.index = top_200_ben_prots.index.map(mapping)\n",
    "top_200_ben_genes.index = top_200_ben_genes.index.map(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "9a448adb-b48a-48ce-813f-9c8691b7779d",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_df_t = env_df.T\n",
    "env_df_t.columns = env_df_t.iloc[0]\n",
    "env_df_t = env_df_t.drop(env_df_t.index[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "eeaa3c84-5d7a-4559-87ed-2668777620a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregate by mean\n",
    "# TODO - why do we have that? only ok if these are replicates\n",
    "top_200_ben_genes = top_200_ben_genes.groupby(top_200_ben_genes.index).mean()\n",
    "top_200_ben_prots = top_200_ben_prots.groupby(top_200_ben_prots.index).mean()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39b7d6d7-ae43-4e40-a5c2-c6ab3671ab55",
   "metadata": {},
   "source": [
    "# Run statistical model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "966d7d98-a9ec-4521-a30d-f46e119d777a",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_sites = top_200_ben_genes.index.intersection(top_200_ben_prots.index).intersection(env_df_t.index)\n",
    "\n",
    "top_200_ben_genes_shared_sites = top_200_ben_genes.loc[common_sites]\n",
    "top_200_ben_prots_shared_sites = top_200_ben_prots.loc[common_sites]\n",
    "env_features_shared_sites = env_df_t.loc[common_sites]\n",
    "\n",
    "full_feature_table = pd.concat([top_200_ben_genes_shared_sites, top_200_ben_prots_shared_sites, env_features_shared_sites], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91dac05d-47e6-47e5-b223-a0cffaf2355d",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_data_t = processed_metabolomic_data.T\n",
    "metabolomic_data_t.columns = metabolomic_data_t.iloc[0]\n",
    "metabolomic_data_t = metabolomic_data_t.drop(metabolomic_data_t.index[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1abb6b14-fda2-496e-b4f5-600af0cad0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_data_t.index = metabolomic_data_t.index.map(mapping)\n",
    "metabolomic_data_shared_sites = metabolomic_data_t.loc[common_sites]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bc4445a-8556-4995-adb5-2c47c5096d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolomic_data_shared_sites = metabolomic_data_shared_sites.groupby(metabolomic_data_shared_sites.index).max()\n",
    "metabolomic_data_shared_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "5c7380fa-0ef8-411b-92be-784cb51a1b3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      545\n",
       "1    17875\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = metabolomic_data_shared_sites.values.flatten()\n",
    "pd.Series(counts).value_counts().sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fe3e3b-2f4a-41bd-a17f-121fd0ea8b42",
   "metadata": {},
   "source": [
    "# Run code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "5ad4604c-6893-4a19-bab7-145a57eabc75",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import StratifiedKFold, cross_val_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "00053492-583b-468f-a36d-56946d222164",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = full_feature_table\n",
    "Y = metabolomic_data_shared_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "e34f9c3e-d969-4af4-ae97-c92bc6166284",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns with NaN: ['Sum_Carotenes', 'Chlorophyll_a']\n"
     ]
    }
   ],
   "source": [
    "# Get column names that contain NaN\n",
    "cols_with_nan = X.columns[X.isnull().any()].tolist()\n",
    "print(f\"Columns with NaN: {cols_with_nan}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "65f169c3-af86-431d-b390-d124add7b010",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MMA-106    3.6627\n",
       "MMA-108    0.9782\n",
       "MMA-114    0.5784\n",
       "MMA-115    1.0038\n",
       "MMA-123    0.4749\n",
       "MMA-124    1.6468\n",
       "MMA-127       NaN\n",
       "MMA-130       NaN\n",
       "MMA-131    0.2128\n",
       "MMA-132    0.2364\n",
       "Name: Chlorophyll_a, dtype: object"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X['Chlorophyll_a']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "cb1f8ec9-6462-4cab-b7ac-c53235a0b91c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MMA-106    0.0819\n",
       "MMA-108     0.039\n",
       "MMA-114    0.0145\n",
       "MMA-115     0.033\n",
       "MMA-123    0.0062\n",
       "MMA-124    0.0546\n",
       "MMA-127       NaN\n",
       "MMA-130       NaN\n",
       "MMA-131    0.0052\n",
       "MMA-132    0.0156\n",
       "Name: Sum_Carotenes, dtype: object"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X['Sum_Carotenes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "3a5d2439-0412-425d-a42a-fc338196bebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "scaler = StandardScaler()\n",
    "X_scaled = pd.DataFrame(scaler.fit_transform(X), index=X.index, columns=X.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "554d3321-3fb7-443a-a728-ec9f68d2ab5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary to store top features per metabolite\n",
    "top_features_per_metabolite = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "9b56a3f1-3525-469f-a469-0c76f4fb9ce2",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Input X contains NaN.\nLogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[143]\u001b[39m\u001b[32m, line 19\u001b[39m\n\u001b[32m     10\u001b[39m clf = LogisticRegression(\n\u001b[32m     11\u001b[39m     penalty=\u001b[33m'\u001b[39m\u001b[33melasticnet\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m     12\u001b[39m     solver=\u001b[33m'\u001b[39m\u001b[33msaga\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     15\u001b[39m     max_iter=\u001b[32m5000\u001b[39m\n\u001b[32m     16\u001b[39m )\n\u001b[32m     18\u001b[39m \u001b[38;5;66;03m# Fit model\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m19\u001b[39m \u001b[43mclf\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_scaled\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     21\u001b[39m \u001b[38;5;66;03m# Get absolute coefficients and sort\u001b[39;00m\n\u001b[32m     22\u001b[39m coefs = pd.Series(np.abs(clf.coef_.ravel()), index=X_scaled.columns)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/base.py:1365\u001b[39m, in \u001b[36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(estimator, *args, **kwargs)\u001b[39m\n\u001b[32m   1358\u001b[39m     estimator._validate_params()\n\u001b[32m   1360\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[32m   1361\u001b[39m     skip_parameter_validation=(\n\u001b[32m   1362\u001b[39m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[32m   1363\u001b[39m     )\n\u001b[32m   1364\u001b[39m ):\n\u001b[32m-> \u001b[39m\u001b[32m1365\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:1247\u001b[39m, in \u001b[36mLogisticRegression.fit\u001b[39m\u001b[34m(self, X, y, sample_weight)\u001b[39m\n\u001b[32m   1244\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1245\u001b[39m     _dtype = [np.float64, np.float32]\n\u001b[32m-> \u001b[39m\u001b[32m1247\u001b[39m X, y = \u001b[43mvalidate_data\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1248\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m   1249\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1250\u001b[39m \u001b[43m    \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1251\u001b[39m \u001b[43m    \u001b[49m\u001b[43maccept_sparse\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcsr\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m   1252\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43m_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1253\u001b[39m \u001b[43m    \u001b[49m\u001b[43morder\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mC\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m   1254\u001b[39m \u001b[43m    \u001b[49m\u001b[43maccept_large_sparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43msolver\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mliblinear\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43msag\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43msaga\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1255\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1256\u001b[39m check_classification_targets(y)\n\u001b[32m   1257\u001b[39m \u001b[38;5;28mself\u001b[39m.classes_ = np.unique(y)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/utils/validation.py:2971\u001b[39m, in \u001b[36mvalidate_data\u001b[39m\u001b[34m(_estimator, X, y, reset, validate_separately, skip_check_array, **check_params)\u001b[39m\n\u001b[32m   2969\u001b[39m         y = check_array(y, input_name=\u001b[33m\"\u001b[39m\u001b[33my\u001b[39m\u001b[33m\"\u001b[39m, **check_y_params)\n\u001b[32m   2970\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m2971\u001b[39m         X, y = \u001b[43mcheck_X_y\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mcheck_params\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   2972\u001b[39m     out = X, y\n\u001b[32m   2974\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m no_val_X \u001b[38;5;129;01mand\u001b[39;00m check_params.get(\u001b[33m\"\u001b[39m\u001b[33mensure_2d\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m):\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/utils/validation.py:1368\u001b[39m, in \u001b[36mcheck_X_y\u001b[39m\u001b[34m(X, y, accept_sparse, accept_large_sparse, dtype, order, copy, force_writeable, force_all_finite, ensure_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, estimator)\u001b[39m\n\u001b[32m   1362\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m   1363\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mestimator_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m requires y to be passed, but the target y is None\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1364\u001b[39m     )\n\u001b[32m   1366\u001b[39m ensure_all_finite = _deprecate_force_all_finite(force_all_finite, ensure_all_finite)\n\u001b[32m-> \u001b[39m\u001b[32m1368\u001b[39m X = \u001b[43mcheck_array\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1369\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1370\u001b[39m \u001b[43m    \u001b[49m\u001b[43maccept_sparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43maccept_sparse\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1371\u001b[39m \u001b[43m    \u001b[49m\u001b[43maccept_large_sparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43maccept_large_sparse\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1372\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1373\u001b[39m \u001b[43m    \u001b[49m\u001b[43morder\u001b[49m\u001b[43m=\u001b[49m\u001b[43morder\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1374\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1375\u001b[39m \u001b[43m    \u001b[49m\u001b[43mforce_writeable\u001b[49m\u001b[43m=\u001b[49m\u001b[43mforce_writeable\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1376\u001b[39m \u001b[43m    \u001b[49m\u001b[43mensure_all_finite\u001b[49m\u001b[43m=\u001b[49m\u001b[43mensure_all_finite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1377\u001b[39m \u001b[43m    \u001b[49m\u001b[43mensure_2d\u001b[49m\u001b[43m=\u001b[49m\u001b[43mensure_2d\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1378\u001b[39m \u001b[43m    \u001b[49m\u001b[43mallow_nd\u001b[49m\u001b[43m=\u001b[49m\u001b[43mallow_nd\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1379\u001b[39m \u001b[43m    \u001b[49m\u001b[43mensure_min_samples\u001b[49m\u001b[43m=\u001b[49m\u001b[43mensure_min_samples\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1380\u001b[39m \u001b[43m    \u001b[49m\u001b[43mensure_min_features\u001b[49m\u001b[43m=\u001b[49m\u001b[43mensure_min_features\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1381\u001b[39m \u001b[43m    \u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m=\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1382\u001b[39m \u001b[43m    \u001b[49m\u001b[43minput_name\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mX\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m   1383\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1385\u001b[39m y = _check_y(y, multi_output=multi_output, y_numeric=y_numeric, estimator=estimator)\n\u001b[32m   1387\u001b[39m check_consistent_length(X, y)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/utils/validation.py:1105\u001b[39m, in \u001b[36mcheck_array\u001b[39m\u001b[34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_writeable, force_all_finite, ensure_all_finite, ensure_non_negative, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator, input_name)\u001b[39m\n\u001b[32m   1099\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m   1100\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFound array with dim \u001b[39m\u001b[38;5;132;01m{\u001b[39;00marray.ndim\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m,\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1101\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m while dim <= 2 is required\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcontext\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1102\u001b[39m     )\n\u001b[32m   1104\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m ensure_all_finite:\n\u001b[32m-> \u001b[39m\u001b[32m1105\u001b[39m     \u001b[43m_assert_all_finite\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1106\u001b[39m \u001b[43m        \u001b[49m\u001b[43marray\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1107\u001b[39m \u001b[43m        \u001b[49m\u001b[43minput_name\u001b[49m\u001b[43m=\u001b[49m\u001b[43minput_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1108\u001b[39m \u001b[43m        \u001b[49m\u001b[43mestimator_name\u001b[49m\u001b[43m=\u001b[49m\u001b[43mestimator_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1109\u001b[39m \u001b[43m        \u001b[49m\u001b[43mallow_nan\u001b[49m\u001b[43m=\u001b[49m\u001b[43mensure_all_finite\u001b[49m\u001b[43m \u001b[49m\u001b[43m==\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mallow-nan\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m   1110\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1112\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m copy:\n\u001b[32m   1113\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m _is_numpy_namespace(xp):\n\u001b[32m   1114\u001b[39m         \u001b[38;5;66;03m# only make a copy if `array` and `array_orig` may share memory`\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/utils/validation.py:120\u001b[39m, in \u001b[36m_assert_all_finite\u001b[39m\u001b[34m(X, allow_nan, msg_dtype, estimator_name, input_name)\u001b[39m\n\u001b[32m    117\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m first_pass_isfinite:\n\u001b[32m    118\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m120\u001b[39m \u001b[43m_assert_all_finite_element_wise\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    121\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    122\u001b[39m \u001b[43m    \u001b[49m\u001b[43mxp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mxp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    123\u001b[39m \u001b[43m    \u001b[49m\u001b[43mallow_nan\u001b[49m\u001b[43m=\u001b[49m\u001b[43mallow_nan\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    124\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmsg_dtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmsg_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    125\u001b[39m \u001b[43m    \u001b[49m\u001b[43mestimator_name\u001b[49m\u001b[43m=\u001b[49m\u001b[43mestimator_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    126\u001b[39m \u001b[43m    \u001b[49m\u001b[43minput_name\u001b[49m\u001b[43m=\u001b[49m\u001b[43minput_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    127\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/conda/lib/python3.13/site-packages/sklearn/utils/validation.py:169\u001b[39m, in \u001b[36m_assert_all_finite_element_wise\u001b[39m\u001b[34m(X, xp, allow_nan, msg_dtype, estimator_name, input_name)\u001b[39m\n\u001b[32m    152\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m estimator_name \u001b[38;5;129;01mand\u001b[39;00m input_name == \u001b[33m\"\u001b[39m\u001b[33mX\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m has_nan_error:\n\u001b[32m    153\u001b[39m     \u001b[38;5;66;03m# Improve the error message on how to handle missing values in\u001b[39;00m\n\u001b[32m    154\u001b[39m     \u001b[38;5;66;03m# scikit-learn.\u001b[39;00m\n\u001b[32m    155\u001b[39m     msg_err += (\n\u001b[32m    156\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mestimator_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m does not accept missing values\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    157\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33m encoded as NaN natively. For supervised learning, you might want\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   (...)\u001b[39m\u001b[32m    167\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33m#estimators-that-handle-nan-values\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    168\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m169\u001b[39m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg_err)\n",
      "\u001b[31mValueError\u001b[39m: Input X contains NaN.\nLogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values"
     ]
    }
   ],
   "source": [
    "# Loop over each metabolite\n",
    "for metabolite in Y.columns:\n",
    "    y = Y[metabolite].values\n",
    "    \n",
    "    # Skip metabolites that are all 0 or all 1\n",
    "    if len(np.unique(y)) < 2:\n",
    "        continue\n",
    "    \n",
    "    # ElasticNet logistic regression\n",
    "    clf = LogisticRegression(\n",
    "        penalty='elasticnet',\n",
    "        solver='saga',\n",
    "        l1_ratio=0.8,   # adjust between 0 (Ridge) and 1 (Lasso)\n",
    "        C=1.0,          # inverse regularization strength\n",
    "        max_iter=5000\n",
    "    )\n",
    "    \n",
    "    # Fit model\n",
    "    clf.fit(X_scaled, y)\n",
    "    \n",
    "    # Get absolute coefficients and sort\n",
    "    coefs = pd.Series(np.abs(clf.coef_.ravel()), index=X_scaled.columns)\n",
    "    top_features = coefs.sort_values(ascending=False).head(10).index.tolist()\n",
    "    \n",
    "    top_features_per_metabolite[metabolite] = top_features"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
